url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw2_P1/Parser.java

```diff
@@ -0,0 +1,182 @@
+package hw2_P1;
+
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.LinkedList;
+
+import javax.swing.text.MutableAttributeSet;
+import javax.swing.text.html.HTML;
+import javax.swing.text.html.HTMLEditorKit;
+import javax.swing.text.html.parser.ParserDelegator;
+
+
+/**
+ * parses pages to pull out links, does some cleanup
+ * TODO: cleanup is faulty, doesn't work with wikipedia
+ * @author tim tregubov
+ * cs44 w10 hw2 p1
+ *
+ */
+public class Parser
+{
+	
+	ParserDelegator parser;  // parse object for html
+	HTMLEditorKit.ParserCallback callback;  //callback on tags
+	ArrayList<String> list;  //temporary store of links
+	
+	public Parser()
+	{
+		parser = new ParserDelegator();
+		callback = new ParseTags();
+		list = new ArrayList<String>();
+	}
+	
+	
+	
+	/**
+	 * gets all the url references in the page
+	 * and fully qualifies them before returning
+	 * @param geturl
+	 * @return
+	 */
+	public LinkedList<String> getAnchorsInPage(String geturl, String target)
+	{
+		list = new ArrayList<String>(); //clean out list
+		LinkedList<String> nlist = new LinkedList<String>(); //for modified urls
+		
+
+		try
+		{
+			Thread.sleep(200);
+
+			URL url = null;
+			
+			try
+			{
+				url = new URL(geturl);
+			} catch (MalformedURLException e)
+			{
+				System.out.println("	bad url: " + geturl);
+			}
+			
+			BufferedReader reader = null;
+			
+			try
+			{
+				reader = new BufferedReader(new InputStreamReader(url.openStream()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *reader*. Currently, there are execution paths that do not contain closure statements, for example, when ParserDelegator.parse throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).